src/libostree/ostree-sysroot.c

```diff
@@ -251,9 +266,46 @@ ostree_sysroot_set_mount_namespace_in_use (OstreeSysroot *self)
   /* Must be before we're loaded, as otherwise we'd have to close/reopen all our
      fds, e.g. the repo */
   g_return_if_fail (self->loadstate < OSTREE_SYSROOT_LOAD_STATE_LOADED);
+  g_autoptr (GError) local_error = NULL;
+  g_assert (!is_in_root_mount_namespace (NULL, &local_error));
+  g_assert (local_error == NULL);
   self->mount_namespace_in_use = TRUE;
 }
 
+gboolean
+_ostree_sysroot_enter_mount_namespace (OstreeSysroot *self, GCancellable *cancellable,
+                                       GError **error)
+{
+  /* Do nothing if we're already in mount namespace */
+  if (self->mount_namespace_in_use)
+    return TRUE;
+
+  /* Do nothing if we're not privileged */
+  if (!ot_util_process_privileged ())
+    return TRUE;
+
+  /* We also assume operating on non-booted roots won't have a readonly sysroot */
+  if (!self->root_is_ostree_booted)
+    return TRUE;
+
+  // If the mount namespaces are the same, we need to unshare().
+  if (is_in_root_mount_namespace (cancellable, error))
+    {
+      if (unshare (CLONE_NEWNS) < 0)
+        return glnx_throw_errno_prefix (error, "Failed to invoke unshare(CLONE_NEWNS)");
+
+      /* Ensure what we do in our mount namespace do not leak to outside */
+      if (mount (NULL, "/", NULL, MS_PRIVATE | MS_REC | MS_SILENT, NULL) < 0)
+        return glnx_throw_errno_prefix (error, "Failed to set the mount propagation to private");
+    }
+  else
+    g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
```

Comment:
This error handling doesn't match the GError conventions. Unfortunately the equivalent of `Result<bool>` as it'd be spelled in Rust needs to be encoded with glib+C as a separate `gboolean *out_somevalue` pointer passed in. See e.g. `_ostree_bootloader_query` or `_ostree_fetcher_request_to_tmpfile_finish` for examples of this pattern.

Reply:
Sorry but I need to point out that you may want to re-pull it. The code you comment on is outdated.